File: question/neetcode150/stack/min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, v: int) -> None:
        if not self.min_stack or v <= self.min_stack[-1]:
            self.min_stack.append(v)
        self.stack.append(v)

    def pop(self) -> None:
        v = self.stack.pop()
        if v == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: question/neetcode150/stack/min-stack.cpp
#include<bits/stdc++.h>

using namespace std;

class MinStack {
public:
    stack<int> s;
    stack<int> ms;

    MinStack() {
        
    }
    
    void push(int val) {
        int v = INT_MAX;
        if (!ms.empty())
            v = ms.top();
        ms.push(min(v,val));
        s.push(val);
    }
    
    void pop() {
        s.pop();
        ms.pop();
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return ms.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */